File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day01NoTimeForATaxicab.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.{East, Facing, FileUtilities, Left, LocationHelper, North, Point, Right, South, Turn, West}

import scala.annotation.tailrec

class Day01NoTimeForATaxicab(filename: String) extends DailyProblem[Int, Int] {
  private def parser(line: String): List[(Turn, Int)] = line.split(", ").map(step => {
    if (step.charAt(0) == 'L') {
      (Left(), step.substring(1).toInt)
    } else {
      (Right(), step.substring(1).toInt)
    }
  }).toList

  private val moves = parser(FileUtilities.readFile(filename)(0))

  private val journey = moves.foldLeft[(Facing, Point, List[Point])]((North(), Point(0, 0), List(Point(0, 0))))((current, step) => {
    val newFacing : Facing = LocationHelper.turn(current._1, step._1)
    val (newPosition : Point, steps : List[Point]) = LocationHelper.steps(current._2, newFacing, step._2, List())
    (newFacing, newPosition, steps ::: current._3)
  })

  def manhattanDistance(current: Point, start: Point): Int = Math.abs(start.y - current.y) + Math.abs(start.x - current.x)

  @tailrec
  private def search(seen: Set[Point], remaining: List[Point]): Point = {
    if (remaining.isEmpty) {
      Point(0, 0)
    } else {
      val current = remaining.head
      if (seen.contains(current)) {
        current
      } else {
        search(seen + current, remaining.tail)
      }
    }
  }

  override val part1Answer: Int = manhattanDistance(journey._2, Point(0, 0))
  override val part2Answer: Int = manhattanDistance(search(Set(), journey._3.reverse), Point(0, 0))
}

File: python-advent-of-code/advent/twenty_sixteen/day01_no_time_for_a_taxicab.py
from advent.utilities.file_reader import FileReader


class Day01NoTimeForATaxicab:
    @staticmethod
    def move_convertor(line):
        return [(bit.strip()[0], int(bit.strip()[1:])) for bit in line.split(',')]

    _left_turn = {
        "North": "West",
        "South": "East",
        "West": "South",
        "East": "North"
    }
    _right_turn = {
        "North": "East",
        "South": "West",
        "West": "North",
        "East": "South"
    }

    _deltas = {
        "North": (0, 1),
        "South": (0, -1),
        "West": (-1, 0),
        "East": (1, 0)
    }

    def __init__(self, filename):
        moves = FileReader.read_file(filename, Day01NoTimeForATaxicab.move_convertor)[0]

        facing = 'North'
        pos = (0, 0)
        first_match = None
        path = []

        for (turn, steps) in moves:
            if turn == 'R':
                facing = self._right_turn[facing]
            if turn == 'L':
                facing = self._left_turn[facing]
            delta = self._deltas[facing]
            for s in range(0, steps):
                pos = (pos[0] + delta[0], pos[1] + delta[1])
                if pos in path and first_match is None:
                    first_match = pos
                path.append(pos)

        self.part_1_answer = self.manhatten_distance((0, 0), pos)
        self.part_2_answer = self.manhatten_distance((0, 0), first_match)

    @staticmethod
    def manhatten_distance(start, end):
        return abs(start[0] - end[0]) + abs(start[1] - end[1])
